File: app/src/main/java/com/mouredev/weeklychallenge2022/Challenge34.dart
/*
 * Reto #34
 * LOS NÚMEROS PERDIDOS
 * Fecha publicación enunciado: 22/08/22
 * Fecha publicación resolución: 29/08/22
 * Dificultad: FÁCIL
 *
 * Enunciado: Dado un array de enteros ordenado y sin repetidos, crea una función que calcule
 * y retorne todos los que faltan entre el mayor y el menor.
 * - Lanza un error si el array de entrada no es correcto.
 *
 * Información adicional:
 * - Usa el canal de nuestro Discord (https://mouredev.com/discord) "🔁reto-semanal"
 *   para preguntas, dudas o prestar ayuda a la comunidad.
 * - Tienes toda la información sobre los retos semanales en
 *   https://retosdeprogramacion.com/semanales2022.
 *
 */

void main() {
  print(lostNumbers([2, 5, 6, 7, 10]));
  print(lostNumbers([-2, 0, 2, 5]));

  try {
    print(lostNumbers([0, 2, 5, 5]));
  } catch (e) {
    print(e);
  }
}

List<int> lostNumbers(List<int> inputList) {
  List<int> lostNumbersList = [];

  for (int index = 0; index < inputList.length - 1; index++) {
    int inputNumber = inputList[index];
    int nextInputNumber = inputList[index + 1];

    // check order and repetitions
    if (inputNumber >= nextInputNumber) {
      throw FormatException(
          "El array de entrada debe estar ordenado de forma ascendente y "
          "no puede tener repeticiones");
    }

    for (int lostNumber = inputNumber + 1; lostNumber < nextInputNumber; lostNumber++) {
      lostNumbersList.add(lostNumber);
    }
  }

  return lostNumbersList;
}

File: app/src/main/java/com/mouredev/weeklychallenge2022/Challenge34.rb
# frozen_string_literal: true

#
# Reto #34
# LOS NÚMEROS PERDIDOS
# Fecha publicación enunciado: 22/08/22
# Fecha publicación resolución: 29/08/22
# Dificultad: FÁCIL
#
# Enunciado: Dado un array de enteros ordenado y sin repetidos, crea una función que calcule
# y retorne todos los que faltan entre el mayor y el menor.
# - Lanza un error si el array de entrada no es correcto.
#
# Información adicional:
# - Usa el canal de nuestro Discord (https://mouredev.com/discord) "🔁reto-semanal"
#   para preguntas, dudas o prestar ayuda a la comunidad.
# - Tienes toda la información sobre los retos semanales en
#   https://retosdeprogramacion.com/semanales2022.
#

# Solución:
# Requerimientos:
# Ruby 2.7.3
#
# Ejecución:
# rubu Challenge34_test.rb

# class that receives and checks input
class Main
  def initialize(arr)
    @arr = arr
  end

  def execute
    if check_input
      lost_numbers = LostNumbers.new(@arr)
      lost_numbers.lost_numbers
    else
      'Not correct input'
    end
  end

  private

  def check_input
    if @arr.instance_of?(Array)
      true if !@arr.empty? && integers? && ordered?
    else
      false
    end
  end

  def ordered?
    @arr.sort.uniq == @arr
  end

  def integers?
    @arr.all? { |i| i.is_a? Integer }
  end
end

# class that receives and returns the lost numbers from the input array
class LostNumbers
  def initialize(arr)
    @arr = arr
  end

  def lost_numbers
    next_num = @arr.first + 1
    index = 1
    resp = []
    while next_num < @arr.last
      @arr[index] != next_num ? resp << next_num : index += 1
      next_num += 1
    end
    resp
  end
end
